### PR TITLE
GDB support and hardware definition using XML

### PR DESCRIPTION
**Pull request by jrb0001**
GDB support:
- run / step works
- breakpoints and read / write watchpoints work
- read / write memory works
- read / write r0 to r15 works
- read / write of floating point registers untested
- read of cpsr works, write of cpsr untested

Basic debugger:
- read / write watchpoints work
- executing a breakpoint or a read watchpoint causes a loop (removing the breakpoint / watchpoint ends the loop)

Hardware definition:
- multiple screens, gpus and keyboards (not really tested)
- moved to an xml file, example:

```
<?xml version="1.0" encoding="UTF-8"?>
<hardwareDefinition>
    <memory>256</memory>
    <rom>bin/rom.rom</rom>
    <devices>
        <fs basepath="bin" writable="false" />
        <screen width="80" height="25" bits="4" />
        <gpu />
        <keyboard />
    </devices>
</hardwareDefinition>
```

General:
- Both debuggers share as much code as possible, a new simple debugger (eg. trace memory access and pc) should be less than 50 lines
- Improved message of BusErrorException to include a simple description and the address
- Simplified command line options, usage:

```
options:
  -trace
    Trace all component invocations.
  -gdb port
    Specifies the port where the gdbserver should listen.
  -gdbverbose
    Specifies the port where the gdbserver should listen.
  -addrinfocmd "command to execute"
    Specifies an external command to use to map instruction addresses to useful information. The command should read hexadecimal addresses one line at a time, and output exactly one line of information for each line of input. (e.g. -addrinfocmd "arm-none-eabi-addr2line -spfe path/to/unstripped_binary.elf")
    The parser that processes the command string is very simple. If you want complex argument escaping, consider making a shell script and executing that.
hwdefinition:
  Path to xml file with the hardware definition
```

### CODE REVIEW

File: src/name/bizna/jarm/PhysicalMemorySpace.java [hunk 1]
```diff
@@ -1,16 +1,42 @@
...
+		private long base, end;
+		private MemoryRegion region;
+
+		public long getBase() {
+			return base;
+		}
+
+		public void setBase(long base) {
+			this.base = base;
+		}
```
Comment (SolraBizna) [inline src/name/bizna/jarm/PhysicalMemorySpace.java]: If the `base` for a `MappedRegion` is changed, the memory map may become unsorted and searches through it may begin to fail. (`base` should probably have originally been `final`.)

File: src/name/bizna/jarm/PhysicalMemorySpace.java [hunk 2]
```diff
@@ -1,16 +1,42 @@
...
+
+		public long getEnd() {
+			return end;
+		}
+
+		public void setEnd(long end) {
+			this.end = end;
+		}
```
Comment (SolraBizna) [inline src/name/bizna/jarm/PhysicalMemorySpace.java]: `end` is merely a cached `region.getRegionSize()`. Rather than letting it be changed directly, update the value when `setRegion(MemoryRegion)` is called.

File: src/name/bizna/jarm/VirtualMemorySpace.java [hunk 3]
```diff
@@ -1,24 +1,32 @@
 package name.bizna.jarm;
 
 public final class VirtualMemorySpace {
-	private PhysicalMemorySpace mem;
+	private final PhysicalMemorySpace mem;
```
Comment (SolraBizna) [inline src/name/bizna/jarm/VirtualMemorySpace.java:4]: If this field is final, then the corresponding field from `CPU` should be final as well.

File: src/name/bizna/jarm/BusErrorException.java [hunk 4]
```diff
@@ -2,4 +2,8 @@
 
 public final class BusErrorException extends Exception {
 	static final long serialVersionUID = 1;
+
+	public BusErrorException(String message) {
+		super(message);
+	}
```
Comment (SolraBizna) [inline src/name/bizna/jarm/BusErrorException.java:33]: I never did get around to making proper ARM exceptions, which would've exposed this information. I thought I'd already added a hack similar to this one to make debugging not awful... but I don't see any sign that I did. (Maybe I'm thinking of something in `*Region`?)

The way I'd recommend doing this would be to make `address` and `accessType` parameters, and building a human-readable message from that on request. That way a debugger can access that information directly and possibly do something with it. See `UnimplementedInstructionException` for an example of the paradigm I mean.

File: src/name/bizna/ocarmsim/BasicDebugger.java [hunk 5]
```diff
@@ -0,0 +1,248 @@
...
+	private final Set<Breakpoint> writeWatchpoints = new HashSet<>();
+
+	public synchronized void reset() {
+		// Reset CPU.
+		cpu.reset(false, true, true);
+
+		// Reset components.
+		((FakeMachine) ((CP3) cpu.getCoprocessor(3)).getMachine()).getComponents().forEach((c) -> {
+			c.reset();
+		});
```
Comment (SolraBizna) [inline src/name/bizna/ocarmsim/BasicDebugger.java]: Use of a lambda here requires Java 8. A number of people are stuck on Java 7 (mostly Mac users).

Maybe something like:

```
for(Component c : ((FakeMachine) ((CP3) cpu.getCoprocessor(3)).getMachine()).getComponents()) {
    c.reset();
}
```

Some similar logic should go into `SimThread` when its execution state is `RESETTING`. (I should have put it there in the first place...)

File: src/name/bizna/ocarmsim/BasicDebugger.java [hunk 6]
```diff
@@ -0,0 +1,248 @@
+package name.bizna.ocarmsim;
+
+import java.awt.Component;
```
Comment (SolraBizna) [inline src/name/bizna/ocarmsim/BasicDebugger.java]: This should almost certainly be `li.cil.oc.api.network.Component`.

File: src/name/bizna/ocarmsim/FakeMachine.java [hunk 7]
```diff
@@ -17,27 +19,28 @@
 import li.cil.oc.api.network.Message;
 import li.cil.oc.api.network.Network;
 import li.cil.oc.api.network.Node;
+import name.bizna.jarm.Debugger;
+import name.bizna.ocarmsim.components.SimComponent;
+import net.minecraft.nbt.NBTTagCompound;
 
 public class FakeMachine implements Machine {
-	
-	public static String computerAddress = "e521fbeb-b67d-434e-9295-40c31361e6f9";
```
Comment (SolraBizna) [inline src/name/bizna/ocarmsim/FakeMachine.java:23]: This does have a reason to exist, I just hadn't implemented it yet. (I think.)

File: src/name/bizna/ocarmsim/FakeMachine.java [hunk 8]
```diff
@@ -160,12 +170,14 @@ public boolean stop() {
 
 	@Override
 	public boolean signal(String name, Object... args) {
-		synchronized(signals) {
-			signals.add(new SimSignal(name, args));
+		if (OCARM.instance.shouldTraceInvocations()) {
+			OCARM.logger.info("signal pushed: %s", name);
 		}
-		synchronized(thread) {
-			thread.notify();
```
Comment (SolraBizna) [inline src/name/bizna/ocarmsim/FakeMachine.java:167]: This `notify` call is necessary, in order to wake up the thread if it's waiting for a signal. (And it should be after `signals.add`.)

File: src/name/bizna/ocarmsim/FakeMachine.java [hunk 9]
```diff
@@ -189,15 +201,14 @@ public Architecture architecture() {
 
 	@Override
 	public Map<String, String> components() {
-		Map<String, String> ret = new HashMap<String, String>();
+		Map<String, String> ret = new HashMap<>();
```
Comment (SolraBizna) [inline src/name/bizna/ocarmsim/FakeMachine.java:204]: What is the reason for this change?

File: src/name/bizna/ocarmsim/FakeMachine.java [hunk 10]
```diff
@@ -263,9 +274,18 @@ public boolean crash(String message) {
 
 	@Override
 	public Signal popSignal() {
-		synchronized(signals) {
-			if(signals.isEmpty()) return null;
-			else return signals.poll();
+		synchronized (signals) {
+			if (signals.isEmpty()) {
+				OCARM.logger.info("signal poped: %s", "{none}");
```
Comment (SolraBizna) [inline src/name/bizna/ocarmsim/FakeMachine.java]: This one should also be protected by `OCARM.instance.shouldTraceInvocations()`. And the proper spelling is `popped`. :)

File: src/name/bizna/ocarmsim/Launcher.java [hunk 11]
```diff
@@ -0,0 +1,100 @@
...
+
+	private static void usage() {
+		System.err.println("Usage: ocarmsim [options] hwdefinition");
+		System.err.println("options:");
+		System.err.println("  -trace");
+		System.err.println("    Trace all component invocations.");
+		System.err.println("  -gdb port");
+		System.err.println("    Specifies the port where the gdbserver should listen.");
+		System.err.println("  -gdbverbose");
+		System.err.println("    Specifies the port where the gdbserver should listen.");
```
Comment (SolraBizna) [inline src/name/bizna/ocarmsim/Launcher.java]: This description does not belong to the option it describes.

File: src/name/bizna/ocarmsim/OCARMSim.java [hunk 12]
```diff
@@ -1,195 +0,0 @@
...
-		machine.addNode(new SimEEPROM(rom, sram));
-		arch.setModule0Size((int)sram.getRegionSize());
-		arch.setSRAMRegion(sram);
-		mem.mapRegion(0x80000000, sram);
-		for(FSSpec spec : filesystems) {
-			machine.addNode(new SimFilesystem(spec.basepath, spec.writable));
-		}
-		new SimUI(screen_tier, machine, cpu, cp3, rom, sram, rams, addrinfocmd);
-	}
-}
```
Comment (SolraBizna) [inline src/name/bizna/ocarmsim/OCARMSim.java:195]: I would prefer it if this means of describing a machine were kept. Having separate main classes for the XML style and the command line style would work for me.

Your new XML method could be the default main class, and then people like me who want the alternate way could do `java -cp ocarmsim-whatever.jar name.bizna.ocarmsim.AltLauncher`.

File: src/name/bizna/ocarmsim/ROMRegion.java [hunk 13]
```diff
@@ -13,8 +12,8 @@
 	// cribbed from JARMArchitecture
 	private static final int ROM_MIN_BIT_DEPTH = 12; // 4K ROM
 	
-	byte[] array;
-	int arrayMask;
+	private final byte[] array;
+	private final int arrayMask;
```
Comment (SolraBizna) [inline src/name/bizna/ocarmsim/ROMRegion.java:16]: It's a little frightening how, even though I _know_ HotSpot optimizes accessors to be just as fast, there were still so many places where my brain _insisted_ that there should be direct access to fields...

File: src/name/bizna/ocarmsim/SRAMRegion.java [hunk 14]
```diff
@@ -80,4 +79,19 @@ public void maybeWriteNVRAM() throws IOException {
...
+		this.sramArray = sramArray;
+	}
+
+	public byte[] getNvramArray() {
+		return nvramArray;
+	}
+
+	public void setNvramArray(byte[] nvramArray) {
+		this.nvramArray = nvramArray;
+	}
```
Comment (SolraBizna) [inline src/name/bizna/ocarmsim/SRAMRegion.java]: Allowing the arrays to change at runtime makes the region size no longer constant. `PhysicalMemorySpace` assumes that regions don't change in size.

**Review by SolraBizna — CHANGES_REQUESTED** [review]
When I first saw this notification, I thought it was an issue and not a pull request. I got as far as thinking "You're asking an _awful_ lot" before realizing my mistake. :)

There are just a few changes I'd like you to make and then I'll be more than happy to merge. (See individual comments.) Beyond that, well done; these are very useful features that you've added. Debugging a live CPU with gdb definitely beats dissecting a core dump, and the possibility of simulating a multi-head OS would make writing one much easier.

File: src/name/bizna/jarm/PhysicalMemorySpace.java (hunk 1 above)
Comment (jrb0001) [reply src/name/bizna/jarm/PhysicalMemorySpace.java]: Oops, fixed.

File: src/name/bizna/jarm/PhysicalMemorySpace.java (hunk 2 above)
Comment (jrb0001) [reply src/name/bizna/jarm/PhysicalMemorySpace.java]: I made PhysicalMemorySpace.end final.

File: src/name/bizna/jarm/VirtualMemorySpace.java (hunk 3 above)
Comment (jrb0001) [reply src/name/bizna/jarm/VirtualMemorySpace.java:4]: Changing the physical memory space on a running CPU doesn't make sense, right?

File: src/name/bizna/jarm/BusErrorException.java (hunk 4 above)
Comment (jrb0001) [reply src/name/bizna/jarm/BusErrorException.java:33]: Yes, good idea. I improved it, what do you think?

File: src/name/bizna/ocarmsim/BasicDebugger.java (hunk 6 above)
Comment (jrb0001) [reply src/name/bizna/ocarmsim/BasicDebugger.java]: No, it's for the GUI. I renamed the method to clarify it.

File: src/name/bizna/ocarmsim/BasicDebugger.java (hunk 5 above)
Comment (jrb0001) [reply src/name/bizna/ocarmsim/BasicDebugger.java]: Mac users are stuck on Java 7? [Java 7 from Oracle is end of life since April 2015](http://www.oracle.com/technetwork/java/eol-135779.html). But ok, I converted all lambdas to Java 7 constructs.

File: src/name/bizna/ocarmsim/FakeMachine.java (hunk 7 above)
Comment (jrb0001) [reply src/name/bizna/ocarmsim/FakeMachine.java:23]: And which reason is it?

File: src/name/bizna/ocarmsim/FakeMachine.java (hunk 8 above)
Comment (jrb0001) [reply src/name/bizna/ocarmsim/FakeMachine.java:167]: The SimThread is replaced by the BasicDebugger so the notify is done there.

File: src/name/bizna/ocarmsim/FakeMachine.java (hunk 9 above)
Comment (jrb0001) [reply src/name/bizna/ocarmsim/FakeMachine.java:204]: Why not? It looks like I couldn't resist to change that. :)

File: src/name/bizna/ocarmsim/FakeMachine.java (hunk 10 above)
Comment (jrb0001) [reply src/name/bizna/ocarmsim/FakeMachine.java]: Fixed.

File: src/name/bizna/ocarmsim/Launcher.java (hunk 11 above)
Comment (jrb0001) [reply src/name/bizna/ocarmsim/Launcher.java]: Copy&Paste error, fixed.

File: src/name/bizna/ocarmsim/OCARMSim.java (hunk 12 above)
Comment (jrb0001) [reply src/name/bizna/ocarmsim/OCARMSim.java:195]: Hmm, is that really needed? I created a wrapper for it which launches the main class after it printed the generated xml to stdout so you can check and / or copy&paste it.

File: src/name/bizna/ocarmsim/ROMRegion.java (hunk 13 above)
Comment (jrb0001) [reply src/name/bizna/ocarmsim/ROMRegion.java:16]: Please tell me a reason why there should be direct access to fields. I don't know any...

File: src/name/bizna/ocarmsim/SRAMRegion.java (hunk 14 above)
Comment (jrb0001) [reply src/name/bizna/ocarmsim/SRAMRegion.java]: Didn't think about that, fixed.

File: src/name/bizna/ocarmsim/BasicDebugger.java (hunk 5 above)
Comment (jrb0001) [reply src/name/bizna/ocarmsim/BasicDebugger.java]: Note: I don't have a JDK 7 installed and I am too lazy to install it. It compiles to target 7 but it might use Java 8 classes.

File: src/name/bizna/jarm/VirtualMemorySpace.java (hunk 3 above)
Comment (SolraBizna) [reply src/name/bizna/jarm/VirtualMemorySpace.java:4]: I can't think of a reason to want to do it.

File: src/name/bizna/ocarmsim/BasicDebugger.java (hunk 5 above)
Comment (SolraBizna) [reply src/name/bizna/ocarmsim/BasicDebugger.java]: I didn't see any other Java 8 classes or features in my review, but I don't have Java 7 myself...

File: src/name/bizna/ocarmsim/FakeMachine.java (hunk 7 above)
Comment (SolraBizna) [reply src/name/bizna/ocarmsim/FakeMachine.java:23]: In my initial review, I missed the part where you made the machine UUID dynamic, and assumed I had simply forgotten to implement `FakeMachine.address()`. You can ignore this comment.

File: src/name/bizna/ocarmsim/FakeMachine.java (hunk 8 above)
Comment (SolraBizna) [reply src/name/bizna/ocarmsim/FakeMachine.java:167]: So the new simulator always has a `Debugger` instance, and that `Debugger` must call `notify()`?

File: src/name/bizna/ocarmsim/FakeMachine.java (hunk 9 above)
Comment (SolraBizna) [reply src/name/bizna/ocarmsim/FakeMachine.java:204]: My knowledge of Java generics is pretty poor, but doesn't this create a warning where there wasn't one before?

File: src/name/bizna/ocarmsim/OCARMSim.java (hunk 12 above)
Comment (SolraBizna) [reply src/name/bizna/ocarmsim/OCARMSim.java:195]: Let me put it this way; I'm don't care enough about the Old Way to reject/delay this PR just because it's gone. :)

File: src/name/bizna/ocarmsim/ROMRegion.java (hunk 13 above)
Comment (SolraBizna) [reply src/name/bizna/ocarmsim/ROMRegion.java:16]: In Java, there isn't a single good reason. At least, there hasn't been since... I dunno, Java 1.1?

File: src/name/bizna/ocarmsim/FakeMachine.java (hunk 8 above)
Comment (jrb0001) [reply src/name/bizna/ocarmsim/FakeMachine.java:167]: The `Debugger` is optional but it allows to add hooks for instructions and memory access. It does also handle the main loop of the simulation but that can also be done without a `Debugger` instance (eg. running in Open Computers).

The call to `.notify()` depends on how that `Debugger` is implemented. If you look at [BasicDebugger](https://github.com/jrb0001/jarm/blob/master/src/name/bizna/ocarmsim/BasicDebugger.java#L252), it is already done for the subclasses. A single threaded design would not need it.

File: src/name/bizna/ocarmsim/FakeMachine.java (hunk 9 above)
Comment (jrb0001) [reply src/name/bizna/ocarmsim/FakeMachine.java:204]: `new HashMap();` creates a warning because it is a raw type, `new HashMap<>();` does not because it can infer the type and is actually the same as `new HashMap<String, String>();`. The diamond operator exists since Java 7. [Question on stackoverflow.com](http://stackoverflow.com/questions/4166966/what-is-the-point-of-the-diamond-operator-in-java-7)

File: src/name/bizna/ocarmsim/OCARMSim.java (hunk 12 above)
Comment (jrb0001) [reply src/name/bizna/ocarmsim/OCARMSim.java:195]: Maybe there is someone who hates XML and refuses to write even a single line. :) But yes, even he would be happy now.

File: src/name/bizna/ocarmsim/FakeMachine.java (hunk 9 above)
Comment (SolraBizna) [reply src/name/bizna/ocarmsim/FakeMachine.java:204]: That's good to know.

File: src/name/bizna/jarm/BusErrorException.java (hunk 4 above)
Comment (SolraBizna) [reply src/name/bizna/jarm/BusErrorException.java:33]: Sorry it took me so long to review this one. I like the new one. It's a little less tolerant of border cases than I'd like (such as a null `reason`) but it'll be fine until (unless?) "real exceptions" are added.

**Comment by SolraBizna** [comment]
Thanks again for all that hard work. Makes me wish I had more time to fill in the remaining core details. (missing instructions, exceptions, etc.)

**Comment by jrb0001** [comment]
Unfortunately, I can't help you with that because I can't even write ARM assembler. But if you need help on other parts, just open an issue and I will try to fix it as soon as I find some time.
